socorro/processor/processor_app.py

```diff
@@ -241,7 +240,7 @@ def _transform(self, crash_id):
                 self.logger.warning('error in processing or saving crash %s', crash_id)
```

Comment:
We only run the processor with a `PolyCrashStorage`, so I think `exc_value` is always a `PolyStorageWhatever` sequence. However, in the case where it isn't, this code looks like it'll raise `exc_value` twice. We can probably rework this like this:

```
# PolyCrashStorage can throw a PolyStorageError which is a sequence of exceptions
# so we should capture each of those separately.
if isinstance(exc_value, collections.Sequence):
    for exc_info_item in exc_value:
        self._capture_error(exc_info_item, crash_id)
        self.logger.warning('error in processing or saving crash %s', crash_id)
```

We don't need to turn the non-sequence into a sequence and capture it and then raise it to get captured again later.

Does that make sense? I'm not sure this is a big deal because I don't think we'll ever run the processor differently. It probably makes the code a little easier to understand. Maybe. I don't know. Wednesday.

Reply:
It sounds like you are saying the behaviour should be:

* If the exception is a ``PolyStorageError``:
  - Send each contained exception to Sentry via ``.capture_error``
  - Log each error at Warning level
  - (*change*) Do not re-raise
* If the exception is something else:
  - (*change*) Do not send to Sentry (yet?)
  - (*change*) Do not log
  - Let it propagate as normal

Did I understand correctly? We'll need to change a few tests to match this.

Reply:
The PolyStorageError is the special case this needs to handle because it's actually a sequence of errors each of which need to be captured and then the PolyStorageError itself should get re-raised. If it's not a PolyStorageError, then we don't need to do anything special with it--we can just reraise it.

With some tweaks, I think the code should be:

```
       except Exception as exc:
            # PolyCrashStorage can throw a PolyStorageError which is a sequence
            # of exc_info items, so we need to capture each one
            if isinstance(exc, collections.Sequence):
                for exc_info_item in exc:
                    self._capture_error(exc_info_item, crash_id)
                    self.logger.warning('error in processing or saving crash %s', crash_id)

            # Re-raise the original exception with the correct traceback
            raise
```

Reply:
OK, then there is no reason to have a generic exception handler, just look for ``PolyStorageError``:

```
except PolyCrashStorage as poly_error:
    # Capture and log the exceptions raised by storage backends
    for exception in poly_error:
        self._capture_error(exception, crash_id)
        self.logger.warning('error in processing or saving crash %s', crash_id)
    raise
```

and we leave other exceptions uncaptured:

* If the exception is a ``PolyStorageError``:
  - Send each contained exception to Sentry via ``.capture_error``
  - Log each error at Warning level
  - Re-raise the ``PolyStorageError``
* If the exception is something else:
  - (*change*) Do not send to Sentry (yet?)
  - (*change*) Do not log
  - Let it propagate as normal

Reply:
I think that's fine, but I don't know for sure if `PolyStorageError` is the only sequence error. You'd have to skulk through the code to figure that out. It's probably the case.

Reply:
I took a look, this appears to be the only one:

* There are no sequence errors in the [standard library](https://docs.python.org/3.7/library/exceptions.html)
* ElasticSearch exceptions like [TransportError](https://github.com/elastic/elasticsearch-py/blob/206f5e2754b01789facc0d50dac5177444afd2f6/elasticsearch/exceptions.py#L37-L87) are simple derived classes from standard exceptions
* Configman [config exceptions](https://github.com/mozilla/configman/blob/master/configman/config_exceptions.py) and [value source exceptions](https://github.com/mozilla/configman/blob/master/configman/value_sources/source_exceptions.py) are also trivial derived classes from ``Exception``
* Our code doesn't raise any other sequence-derived exceptions, but we do have this interesting code that raises an integer:

https://github.com/mozilla-services/socorro/blob/eb37a3215d992be987c2d3308d4eb42b666efd78/socorro/lib/cache.py#L107-L108

I think it would be safe, and better practice to make this code less generic. I'll add a new commit that changes this code, and the tests around exceptions that aren't `PolyStorageError`.

Reply:
Sounds good! :+1: